Mathlib/GroupTheory/CosetCover.lean

```diff
@@ -0,0 +1,394 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, Richard Copley. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, Richard Copley
+-/
+
+import Mathlib.GroupTheory.Complement
+import Mathlib.LinearAlgebra.Basis.VectorSpace
+
+/-! # Lemma of B. H. Neumann on coverings of a group by cosets.
+
+Let the group $G$ be the union of finitely many, let us say $n$, left cosets
+of subgroups $C₁$, $C₂$, ..., $Cₙ$: $$ G = ⋃_{i = 1}^n C_i g_i. $$
+
+* `Subgroup.exists_finiteIndex_of_leftCoset_cover`
+  at least one subgroup $C_i$ has finite index in $G$.
+
+* `Subgroup.leftCoset_cover_filter_FiniteIndex`
+  the cosets of subgroups of infinite index may be omitted from the covering.
+
+* `Subgroup.exists_index_le_card_of_leftCoset_cover` :
+  the index of (at least) one of these subgroups does not exceed $n$.
+
+* `Subgroup.one_le_sum_inv_index_of_leftCoset_cover` :
+  the sum of the inverses of the indexes of the $C_i$ is greater than or equal to 1.
+
+* `Subgroup.pairwiseDisjoint_leftCoset_cover_of_sum_inv_index_eq_one`
+  If the sum of the inverses of the indexes of the subgroups $C_i$ is equal to 1,
+  then the cosets of the subgroups of finite index are pairwise disjoint.
+
+A corollary of `Subgroup.exists_finiteIndex_of_leftCoset_cover` is:
+
+* `Subspace.union_ne_univ_of_lt_top` :
+  a vector space over an infinite field cannot be a finite union of proper subspaces.
+
+This can be used to show that an algebraic extension of fields is determined by the
+set of all minimal polynomials (not proved here).
+
+[1] [Neumann-1954], *Groups Covered By Permutable Subsets*, Lemma 4.1
+[2] <https://mathoverflow.net/a/17398/3332>
+[3] <http://alpha.math.uga.edu/~pete/Neumann54.pdf>
+
+-/
+
+open scoped Pointwise BigOperators
+
+section leftCoset_cover_const
+
+variable {G : Type*} [Group G]
+
+@[to_additive]
+theorem Subgroup.exists_leftTransversal_of_FiniteIndex
```

Comment:
Please use a `namespace Subgroup` to shorten the names of the lemmas up to line 338.

Reply:
Done!